### PR TITLE
WIP 1797 timeout regression

### PR DESCRIPTION
**Pull request by MRigal**
## What do these changes do?

Trying first to write a solid failing test case to illustrate #1797 
Not stable for now, but still maybe worth looking at the results

## Are there changes in behavior for the user?

Nope

## Related issue number

Not good to merge for now, aiming at #1797 

## Checklist

- [x] I think the code is well written
- [x] Unit tests for the changes exist
- [x] Documentation reflects the changes
- [x] If you provide code modification, please add yourself to `CONTRIBUTORS.txt`
  * The format is &lt;Name&gt; &lt;Surname&gt;.
  * Please keep alphabetical order, the file is sorted by names. 
- [ ] Add a new entry to `CHANGES.rst`
  * Choose any open position to avoid merge conflicts with other PRs.
  * Add a link to the issue you are fixing (if any) using `#issue_number` format at the end of changelog message. Use Pull Request number if there are no issues for PR or PR covers the issue only partially.

### CODE REVIEW

**Comment by fafhrd91** [comment]
did you read my comment on actual issue?

could you try to shield actual session.get() call.

**Comment by MRigal** [comment]
I'm actually not very sure what you mean with it... Would you like to submit a PR to my branch or just write down code-wise what you are meaning?

**Comment by MRigal** [comment]
OK, I am not able to write a proper test case simulating the full functionality of a proxy, but I could put in evidence on my server that I had to go through a proxy to be able to reproduce the problem. Therefore I updated the issue. Still, downgrading to 1.3.x simply fixes the problem

**Comment by MRigal** [comment]
The report of the travis 3.6 is pretty complete (https://travis-ci.org/aio-libs/aiohttp/jobs/225670002):
```
       6 failed
         - tests/test_py35/test_client.py:104 test_timeout_using_conn_limit[pyloop-10-None]
         - tests/test_py35/test_client.py:104 test_timeout_using_conn_limit[pyloop-3-1000]
         - tests/test_py35/test_client.py:104 test_timeout_using_conn_limit[pyloop-10-1000]
         - tests/test_py35/test_client.py:104 test_timeout_using_conn_limit[uvloop-10-None]
         - tests/test_py35/test_client.py:104 test_timeout_using_conn_limit[uvloop-3-1000]
         - tests/test_py35/test_client.py:104 test_timeout_using_conn_limit[uvloop-10-1000]
```
While applying the same tests on the 1.3 branch will only cause the `*loop-*-1000` to fail... Which shows a behaviour difference (regression?) when not touching the `limit` parameter on the 2.x branch

**Comment by fafhrd91** [comment]
here is difference how aiohttp handles timeout.

1.3 - timeout get applied only to connection + send request, after that there is no timeout on operation
2.0 - timeout get applied to all operations related to request, connection, send request, get response.

in your test you create all connections at the same time, so aiohttp tries to connect and start timeout at the same time. but remote service can not serve you requests fast enough, so at some point you start to see timeout errors. I start to see timeouts around 450 requests, so aiohttp able to complete about 450 request within first 4 seconds. and that is the reason why you can not reproduce this with local server, it just able to serve all your requests within 4 seconds.

also `asyncio.gather` does not protect you from exceptions, (shield() does)
lets say you try to gather 100 requests (coroutines), but if one of those raises exception (TimeoutError in your case) it cancels `asyncio.gather()` operation. for crawler case you can not really just send request and wait response, you have to handle exceptions from requests, bad responses, etc and return result.
generally, you have to "shield" request if you want to use it with `gather()`

**Comment by fafhrd91** [comment]
@MRigal ping

**Comment by fafhrd91** [comment]
@MRigal ?

File: tests/test_py35/test_client.py [hunk 1]
```diff
@@ -99,3 +99,42 @@
...
+    conn_kwargs = dict(loop=loop)
+    if conn_limit is not None:
+        conn_kwargs['limit'] = conn_limit
+    connector = aiohttp.TCPConnector(**conn_kwargs)
+    async with aiohttp.ClientSession(connector=connector,
+                                     loop=loop) as session:
+        for i in range(connector._limit * rows_multiplier):
+            rows.append(asyncio.ensure_future(
+                read_resp(session, server), loop=loop))
+        await asyncio.gather(*rows)
```
Comment (AraHaan) [inline tests/test_py35/test_client.py:140]: Use asyncio.shield() here. This is what @fafhrd91 also suggested above.

**Pull request closed without merging**